### PR TITLE
Cache template directory in init command

### PR DESCRIPTION
**Pull request by hpmellema**
*Description of changes:*
Caches a copy of the smithy init template directory in the default system temp directory. This greatly speeds up subsequent executions of both the list and create template subcommands of the `smithy init`. 
Previously, creating a template could take up to 6 seconds. With this change that time is down to ~0.3sec after the initial caching of the template repo.

---

*Example Executions*
Create a Template
cli

List Templates 
list

Note: the gif recorder used here adds additional lag so these do not accurately represent the time these commands take to execute.
___

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-cli/src/main/java/software/amazon/smithy/cli/commands/InitCommand.java [hunk 1]
```diff
@@ -86,32 +90,100 @@ private int run(Arguments arguments, Env env) {
...
         StandardOptions standardOptions = arguments.getReceiver(StandardOptions.class);
 
-        try {
-            Path root = Paths.get(".");
-            Path temp = Files.createTempDirectory("temp");
+        boolean isLocalRepo = isLocalRepo(options.repositoryUrl);
+        Path templateRepoDirPath = getTemplateRepoDirPath(options.repositoryUrl, isLocalRepo);
+
+        // If the cache directory does not exist, create it
+        if (!Files.exists(templateRepoDirPath)) {
```
Comment (AndrewFossAWS) [inline smithy-cli/src/main/java/software/amazon/smithy/cli/commands/InitCommand.java:97]: Does it make sense to add a param to allow user to invalidate the cache?

Comment (mtdowling) [reply smithy-cli/src/main/java/software/amazon/smithy/cli/commands/InitCommand.java:97]: `smithy clean` does that

**Comment by mtdowling** [comment]
rather than cache to the temp directory, I suggest caching to the build directory, allowing `smithy clean` to be used to clear any caches.

**Review by mtdowling — CHANGES_REQUESTED** [review]
Is it possible to add integration tests for the added functionality of the init command?

**Review by mtdowling — APPROVED** [review]

**Comment by hpmellema** [comment]
For some reason the file lock on the `.git` pack files on the windows runner do not seem to be released (possibly due to the windows git file watcher?). I am working on trying to find the root cause and a fix, but in the mean time this PR will have to wait for that fix.

**Pull request merged**